spokestack/asr/cloud_client.py

```diff
@@ -0,0 +1,206 @@
+"""
+This module contains the websocket logic used to communicate with
+Spokestack's cloud-based ASR service.
+"""
+import base64
+import hashlib
+import hmac
+import json
+from typing import Any, Dict, List
+
+import numpy as np  # type: ignore
+from websocket import WebSocket  # type: ignore
+
+
+class CloudClient:
+    """ Spokestack client for cloud based speech to text
+
+        Args:
+            socket_url (str): url for socket connection
+            key_id (str): identity from spokestack api credentials
+            key_secret (str): secret key from spokestack api credentials
+            audio_format (str): format of input audio
+            sample_rate (int): audio sample rate (kHz)
+            language (str): language for recognition
+            limit (int): Limit of messages per api response
+            idle_timeout (Any): Time before client timeout. Defaults to None
+    """
+
+    def __init__(
+        self,
+        socket_url: str,
+        key_id: str,
+        key_secret: str,
+        audio_format: str = "PCM16LE",
+        sample_rate: int = 16000,
+        language: str = "en",
+        limit: int = 10,
+        idle_timeout: Any = None,
+    ) -> None:
+
+        self._body: str = json.dumps(
+            {
+                "format": audio_format,
+                "rate": sample_rate,
+                "language": language,
+                "limit": limit,
+            }
+        )
+        self._socket_url: str = socket_url
+        self._key_id: str = key_id
+        self._key: bytes = key_secret.encode("utf-8")
+        signature = hmac.new(
+            self._key, self._body.encode("utf-8"), hashlib.sha256
+        ).digest()
+        self._signature = base64.b64encode(signature).decode("utf-8")
+        self._socket: Any = None
+
+        self._response: Dict[str, Any] = {
+            "error": None,
+            "final": False,
+            "hypotheses": [],
+            "status": None,
+        }
+        self._sample_rate: int = sample_rate
+        self._is_final: bool = False
+        self._idle_timeout = idle_timeout
+        self._idle_count: int = 0
+
+    def __call__(self, audio: np.ndarray, n_best: int = 1) -> List[str]:
+        """ Audio to text interface for the cloud client
+
+        Args:
+            audio (np.ndarray): np.int16 array of audio
+            n_best (int): number of predictions to return
+
+        Returns: n_best list of transcripts
+
+        """
+
+        chunk_size = 2 * int(0.01 * self._sample_rate)
+        self.connect()
+        self.initialize()
+
+        for i in range(0, len(audio), chunk_size):
+            frame = audio[i:][:chunk_size]
+            self.send(frame)
+            self.receive()
+
+        self.end()
+        while not self._is_final:
+            if self._idle_timeout and (self._idle_count > self._idle_timeout):
```

Comment:
idle checks don't make sense for bulk recognition, just wait until `is_final` or an error occurs